src/System.Buffers.Primitives/System/Internal/ManagedBufferPool.cs

```diff
@@ -89,7 +90,10 @@ public override void Retain()
             public override void Release()
             {
                 if (!IsRetained) BufferPrimitivesThrowHelper.ThrowInvalidOperationException();
-                Interlocked.Decrement(ref _referenceCount);
+                if(Interlocked.Decrement(ref _referenceCount) <= 0)
```

Comment:
If there was ever a case where _referenceCount could be < 0, it means that someone called Release without having previously called Retain. Assume that bug exists, what happens when your dispose method calling the ArrayPool return method twice with the same array reference?